src/namespace.rs

```diff
@@ -0,0 +1,167 @@
+    // it in the `Drop` call.
+    oldcwd: PathBuf,
+}
+
+fn get_nspid(pid: Pid) -> Result<(Pid, Pid), Error> {
+    let fname = format!("/proc/{pid}/status");
+    let file = File::open(&fname).context("faild to open `{fname}`")?;
+    let reader = BufReader::new(file);
+    let (mut tgid, mut nstgid) = (pid, pid);
+    let mut found = false;
+
+    for line in reader.lines() {
+        match line {
+            Ok(line) => {
+                /* Use tgid if CONFIG_PID_NS is not defined. */
+                if line.contains("Tgid:") {
+                    if let Some(num) = line
+                        .split("Tgid:")
+                        .last()
+                        .and_then(|s| s.split_whitespace().next_back())
+                    {
+                        let id = Pid::from_str(num).map_err(|e| {
+                            io::Error::new(io::ErrorKind::InvalidInput, format!("{e}"))
+                        })?;
+                        tgid = id;
+                        nstgid = id;
+                        found = true;
+                    }
+                }
+                if line.contains("NStgid:") {
+                    if let Some(num) = line
+                        .split("NStgid:")
+                        .last()
+                        .and_then(|s| s.split_whitespace().next_back())
+                    {
+                        nstgid = Pid::from_str(num).map_err(|e| {
+                            io::Error::new(io::ErrorKind::InvalidInput, format!("{e}"))
+                        })?;
+                        break;
+                    }
+                }
+            }
+            Err(e) => return Err(e.into()),
+        }
+    }
+
+    if !found {
+        unreachable!("{}", format!("failed to get Tgid/NStgid from {fname}"));
+    }
+    Ok((tgid, nstgid))
+}
+
+impl NsInfo {
+    pub(crate) fn new(pid: Pid) -> Result<Self, Error> {
+        let old_stat_path = "/proc/self/ns/mnt";
+        let new_stat_path = format!("/proc/{pid}/ns/mnt");
+        let old_stat = fs::metadata(old_stat_path).context("failed to stat `/proc/self/ns/mnt`")?;
+        let new_stat =
+            fs::metadata(&new_stat_path).context("failed to stat `/proc/{pid}/ns/mnt`")?;
+        let oldns = File::open(old_stat_path).context("failed to open `/proc/self/ns/mnt`")?;
+        let oldcwd = current_dir().context("failed to get current work dir")?;
+        let (tgid, nstgid) = get_nspid(pid).context("failed to get nspid for pid {pid}")?;
+        let need_setns = old_stat.ino() != new_stat.ino();
+        let mntns_path = if need_setns {
+            Some(PathBuf::from(new_stat_path))
+        } else {
+            None
+        };
+        Ok(Self {
+            tgid,
+            nstgid,
+            need_setns,
+            mntns_path,
+            oldns,
+            oldcwd,
+        })
+    }
+
+    pub(crate) fn enter_mntns(&self) -> Result<(), Error> {
```

Comment:
How about we "enter" directly in the constructor? The current code will likely error out in the destructor if `new` was called but no `enter_mntns` ever done.